[28]: actualize README file (#29)
RonasIT/laravel-exponent-push-notifications: README.md

Reviewer: let's remove the config file content, we'll describe configs my the native Laravel way, not in the Readme

[![Coverage Status](https://coveralls.io/repos/github/RonasIT/laravel-exponent-push-notifications/badge.svg?branch=master)](https://coveralls.io/github/RonasIT/laravel-exponent-push-notifications?branch=master)
[![Total Downloads](https://img.shields.io/packagist/dt/ronasit/laravel-exponent-push-notifications.svg?style=flat-square)](https://packagist.org/packages/ronasit/laravel-exponent-push-notifications)

> This package provides an Expo push notification channel for Laravel, enabling you to send notifications to mobile devices via the [Expo SDK](https://docs.expo.dev/push-notifications/overview/).

Reviewer: ```suggestion
This package provides an Expo push notification channel for Laravel, allowing you to send notifications to mobile devices via the [Expo SDK](https://docs.expo.dev/push-notifications/overview).
```

Author: fixed

[![Coverage Status](https://coveralls.io/repos/github/RonasIT/laravel-exponent-push-notifications/badge.svg?branch=master)](https://coveralls.io/github/RonasIT/laravel-exponent-push-notifications?branch=master)
[![Total Downloads](https://img.shields.io/packagist/dt/ronasit/laravel-exponent-push-notifications.svg?style=flat-square)](https://packagist.org/packages/ronasit/laravel-exponent-push-notifications)

This package provides an Expo push notification channel for Laravel, enabling you to send notifications to mobile devices via the [Expo SDK](https://docs.expo.dev/push-notifications/overview/).

## Contents

-   [Installation](#installation)
-   [Usage](#usage)

>     -   [Available Message methods](#available-message-methods)

Reviewer: ```suggestion
    -   [Message customization](#available-message-methods)
```

>     -   [Managing Recipients](#managing-recipients)

Reviewer: let's also move it to the start of the usage list

```suggestion
    -   [Push tokens API](#managing-recipients)
```

> You can install the package via composer:

Reviewer: ```suggestion
Install the package via composer:
```

    ...
    NotificationChannels\ExpoPushNotifications\ExpoPushNotificationsServiceProvider::class,

> ],

Reviewer: let's remove this block as we have min Laravel version as 11

```

> You can optionally publish the config file with:

Reviewer: ```suggestion
Publish the config file (optionally):
```

```php
return [
    'debug' => env('EXPONENT_PUSH_NOTIFICATION_DEBUG', true), // log requests to Expo via SDK package

    'middleware' => ['auth:api', 'bindings'],

    'interests' => [
        'driver' => env('EXPONENT_PUSH_NOTIFICATION_INTERESTS_STORAGE_DRIVER', 'file'),

        'database' => [
            'events' => [],
            'table_name' => 'exponent_push_notification_interests',
        ],
    ],
];
```

### Storage Drivers

> | Driver | When to use |

Reviewer: ```suggestion
Package provides several drivers to store the push tokens:

| Driver | When to use |
```

|--------|-------------|
| `file` (default) | Simple setups, single-server, no database required |
| `database` | Multi-server deployments, persistent storage, flexible queries |

When using the `database` driver, add the following to your `.env`:

```dotenv
EXPONENT_PUSH_NOTIFICATION_INTERESTS_STORAGE_DRIVER=database

> ```

Reviewer: ```suggestion
Storage driver may be configured via the `EXPONENT_PUSH_NOTIFICATION_INTERESTS_STORAGE_DRIVER` env variable.
```

Then publish and run the migration:

```bash
php artisan vendor:publish --provider="NotificationChannels\ExpoPushNotifications\ExpoPushNotificationsServiceProvider" --tag="migrations"
php artisan migrate
```

> > The migration uses the table name defined in `interests.database.table_name` from the config (`exponent_push_notification_interests` by default).

Reviewer: ```suggestion
#### Database driver

To use the `database` driver need to publish and run package migration:

```bash
php artisan vendor:publish --provider="NotificationChannels\ExpoPushNotifications\ExpoPushNotificationsServiceProvider" --tag="migrations"
php artisan migrate

> [!IMPORTANT]
> The migration uses the table name defined in `interests.database.table_name` config. Default value is `exponent_push_notification_interests`.
```

}
```

### Push tokens API

The package automatically registers the following endpoints for managing device subscriptions:

#### `POST /exponent/devices/subscribe`

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `expo_token` | string | Yes | The Expo push token of the device to subscribe |

#### `POST /exponent/devices/unsubscribe`

| Field | Type | Required | Description |
|-------|------|----------|-------------|

> | `expo_token` | string | No | The Expo push token to remove. If omitted, all tokens for the authenticated user are removed |

Reviewer: ```suggestion
| `expo_token` | string | No | The Expo push token to remove. If omitted, all tokens for the authenticated user will be removed |
```

Author: fixed

The middleware applied to these endpoints is configured via the `middleware` key in the config file.

The package registers all endpoints automatically. To customize their middleware, prefix, or guards, call `Route::expo()` in your routes file — this disables the automatic registration:

```php

> // In routes/api.php

Reviewer: ```suggestion
// routes/api.php
```

Author: fixed

// Use defaults:
Route::expo();

// Custom prefix and middleware:
Route::expo(

>     prefix: 'api/expo', 

Reviewer: let's disable ability to pass prefix and middleware additionally

Author: Yes, it really does look like an unnecessary function, and it's not obvious for the user which mechanism should be used - default chaining or passing arguments. Removed.

    middleware: ['api', 'auth:sanctum'],

> );

Reviewer: let's add more live example

```suggestion
Route::prefix('v{version}')->group(function () {
    Route::middleware('auth_group')->group(function () {
        // other routes

        Route::expo();
    });
});
```

Author: fixed

}
```

### Push tokens API

The package automatically registers two endpoints for managing device subscriptions:

#### `POST /exponent/devices/subscribe`

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `expo_token` | string | Yes | The Expo push token of the device to subscribe |

#### `POST /exponent/devices/unsubscribe`

| Field | Type | Required | Description |
|-------|------|----------|-------------|
| `expo_token` | string | No | The Expo push token to remove. If omitted, all tokens for the authenticated user are removed |

The middleware applied to these endpoints is configured via the `middleware` key in the config file.

By default, both endpoints are registered automatically by the ServiceProvider. Calling `Route::expo()` in your own routes file registers the same routes, giving you full control over middleware, prefixes, and guards:

```php
use NotificationChannels\ExpoPushNotifications\Enum\ExpoRouterActionEnum;

// Register both routes:
Route::expo();

> // Subscribe only:

Reviewer: @vitgrams I not sure but is seems that the registration of the only API will disable:
- default routes
- the second unregistered route

If it correct need to:
- describe this logic in the readme (about disabling default routes)
- remove optional passed endpoints type (https://github.com/RonasIT/laravel-exponent-push-notifications/blob/6a8ae95cb5537ccef4a5c99f55250cc99a02c0c0/src/ExpoRouter.php#L15)

Author: @DenTray You're right — that's exactly how it was supposed to work, but the implementation was incomplete. The flag $isBlockedBaseRoutes was being set but never actually checked anywhere, so default routes always loaded regardless of whether Route::expo() was called.

Here's what I did:
- Fixed the blocking behavior by checking the flag in a booted() callback in the ServiceProvider, which fires after all providers (including the app's RouteServiceProvider) have booted — so the flag is guaranteed to be set before the check runs
- Removed the optional ExpoRouterActionEnum ...$options parameter, as discussed — selective route registration combined with blocking would silently disable the unregistered route, which is unexpected behavior. Route::expo() now always registers both routes
- Added tests covering: default auto-registration, blocking when Route::expo() is called, and custom prefix/middleware via the macro

Could you take a look and share your thoughts?

### Available Message methods

> -   `title('Your title')`: Sets the notification title.

Reviewer: ```suggestion
`ExpoMessage` may be customized using the next methods: 

-   `title`: Sets the notification title.
```

Author: @DenTray Removed method args description from readme

-   `body('Your message')`: Sets the notification body.
-   `enableSound()`: Enables the notification sound.
-   `disableSound()`: Mutes the notification sound.
-   `badge(1)`: Sets the badge count on the app icon.
-   `setTtl(60)`: Sets the time to live in seconds.
-   `setJsonData(['key' => 'value'])`: Sets additional payload data (JSON string or array). Throws `CouldNotCreateMessage` on invalid JSON.
-   `setChannelId('default')`: Sets the Android notification channel ID.
-   `priority(PriorityEnum::High)`: Sets the notification priority. Available: `PriorityEnum::Default`, `::Normal`, `::High`.

> -   `setInterruptionLevel(InterruptionLevelEnum::Active)`: Sets the iOS interruption level. Available: `InterruptionLevelEnum::Active`, `::Critical`, `::Passive`, `::TimeSensitive`.

Reviewer: let' use table for methods list with the next fields:
- method
- description
- notes (e.g. Android only, default value, available values, etc.)

For example, to send a high-priority time-sensitive notification on iOS:

```php
use NotificationChannels\ExpoPushNotifications\ExpoMessage;
use NotificationChannels\ExpoPushNotifications\Enums\PriorityEnum;
use NotificationChannels\ExpoPushNotifications\Enums\InterruptionLevelEnum;

ExpoMessage::create()
    ->title('Important!')
    ->body('This is a critical alert.')
    ->priority(PriorityEnum::High)
    ->setInterruptionLevel(InterruptionLevelEnum::TimeSensitive);
```

### Managing Recipients

The package automatically registers two endpoints for managing device subscriptions:

```
POST /exponent/devices/subscribe
POST /exponent/devices/unsubscribe
```

**Request body:**

```json
{
    "expo_token": "ExponentPushToken[xxxxxxxxxxxxxxxxxxxxxx]"
}

> ```

Reviewer: Both endpoints require the `expo_token` body field

Author: @DenTray No, it’s not required in the `/unsubscribe` endpoint - it uses the `sometimes` rule.

> ```

Author: @DenTray Fixed, added the separate tables to describe endpoints params

The middleware applied to these endpoints is configured via the `middleware` key in the config file.

### Routing a message

> By default, the Expo "interest" that a message is sent to is derived from the notifiable model using the `{ClassName}.{id}` convention (e.g. `App\Models\User.1`).

Author: @DenTray Fixed

You can override this by adding a `routeNotificationForExpoPushNotifications()` method to your notifiable model:

```php
use Illuminate\Notifications\Notifiable;

class User extends Model
{
    use Notifiable;

    public function routeNotificationForExpoPushNotifications(): string
    {
        return 'App.User.' . $this->id;
    }
}
```

> ### Custom Routes

Reviewer: let's union this block with the routes registration

> By default, both endpoints are registered automatically by the ServiceProvider. Calling `Route::expo()` in your own routes file disables the default route registration, giving you full control over middleware, prefixes, and guards:

Author: @DenTray Fixed
